pkg/binpack/minimal_fragmentation.go

```diff
@@ -0,0 +1,173 @@
+	"gopkg.in/inf.v0"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// MinimalFragmentation is a SparkBinPackFunction that tries to put the driver pod on the first possible node with
+// enough capacity, and then tries to pack executors onto as few nodes as possible.
+var MinimalFragmentation = SparkBinPackFunction(func(
+	ctx context.Context,
+	driverResources, executorResources *resources.Resources,
+	executorCount int,
+	driverNodePriorityOrder, executorNodePriorityOrder []string,
+	nodesSchedulingMetadata resources.NodeGroupSchedulingMetadata) (string, []string, bool) {
+	return SparkBinPack(ctx, driverResources, executorResources, executorCount, driverNodePriorityOrder, executorNodePriorityOrder, nodesSchedulingMetadata, minimalFragmentation)
+})
+
+// minimalFragmentation attempts to pack executors onto as few nodes as possible, ideally a single one
+// nodePriorityOrder is still used as a guideline, i.e. if an application can fit on multiple nodes, it will pick
+// the first eligible node according to nodePriorityOrder
+func minimalFragmentation(
+	_ context.Context,
+	executorResources *resources.Resources,
+	executorCount int,
+	nodePriorityOrder []string,
+	nodeGroupSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	reservedResources resources.NodeGroupResources) ([]string, bool) {
+	executorNodes := make([]string, 0, executorCount)
+	if executorCount == 0 {
+		return executorNodes, true
+	}
+
+	nodeCapacities := getNodeCapacities(nodePriorityOrder, nodeGroupSchedulingMetadata, reservedResources, executorResources)
+	sort.SliceStable(nodeCapacities, func(i, j int) bool {
+		return nodeCapacities[i].capacity < nodeCapacities[j].capacity
+	})
+
+	// as long as we have nodes where we could schedule executors
+	for len(nodeCapacities) > 0 {
+		// pick the first node that could fit all the executors (if there's one)
+		position := sort.Search(len(nodeCapacities), func(i int) bool {
+			return nodeCapacities[i].capacity >= executorCount
+		})
+
+		if position != len(nodeCapacities) {
+			// we found a node that has the required capacity, schedule everything there and we're done
+			return append(executorNodes, repeat(nodeCapacities[position].nodeName, executorCount)...), true
+		}
+
+		// we will need multiple nodes for scheduling, thus we'll try to schedule executors on nodes with the most capacity
+		maxCapacity := nodeCapacities[len(nodeCapacities)-1].capacity
+		position = sort.Search(len(nodeCapacities), func(i int) bool {
+			return nodeCapacities[i].capacity >= maxCapacity
+		})
+
+		// we can skip the check on position since we know at least one node will be found
+		executorNodes = append(executorNodes, repeat(nodeCapacities[position].nodeName, maxCapacity)...)
+		executorCount -= maxCapacity
+
+		// at least on paper, this is somewhat inefficient (but more readable than the alternative)
+		// to be optimized later if needed
+		nodeCapacities = append(nodeCapacities[:position], nodeCapacities[position+1:]...)
```

Comment:
I could be misreading this but I think we can replace all of this with a simple loop through `nodeCapacities` descending which decrements the number of remaining executors by the capacity of the current node and builds up the node names that way, this personally seems simpler to me, what do you think?